controller/snapshot_controller.go

```diff
@@ -0,0 +1,543 @@
+	if !ok {
+		deletedState, ok := oldObj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("received unexpected obj: %#v", oldObj))
+			return
+		}
+		// use the last known state, to enqueue, dependent objects
+		oldEngine, ok = deletedState.Obj.(*longhorn.Engine)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object: %#v", deletedState.Obj))
+			return
+		}
+	}
+
+	curEngine, ok := curObj.(*longhorn.Engine)
+	if !ok {
+		deletedState, ok := curObj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("received unexpected obj: %#v", curObj))
+			return
+		}
+		// use the last known state, to enqueue, dependent objects
+		curEngine, ok = deletedState.Obj.(*longhorn.Engine)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object: %#v", deletedState.Obj))
+			return
+		}
+	}
+
+	needEnqueueSnapshots := curEngine.Status.CurrentState != oldEngine.Status.CurrentState ||
+		!reflect.DeepEqual(curEngine.Status.PurgeStatus, oldEngine.Status.PurgeStatus) ||
+		!reflect.DeepEqual(curEngine.Status.Snapshots, oldEngine.Status.Snapshots)
+
+	if !needEnqueueSnapshots {
+		return
+	}
+
+	snapshots, err := sc.ds.ListVolumeSnapshotsRO(curEngine.Spec.VolumeName)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to list snapshots for volume %v when enqueuing engine %v: %v", curEngine.Spec.VolumeName, oldEngine.Name, err))
+		return
+	}
+	for _, snap := range snapshots {
+		sc.enqueueSnapshot(snap)
+	}
+
+	return
+}
+
+func (sc *SnapshotController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer sc.queue.ShutDown()
+
+	sc.logger.Info("Start Longhorn Snapshot Controller")
+	defer sc.logger.Info("Shutting down Longhorn Snapshot Controller")
+
+	if !cache.WaitForNamedCacheSync(sc.name, stopCh, sc.cacheSyncs...) {
+		return
+	}
+
+	// Init and Start the monitoring
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (sc *SnapshotController) worker() {
+	for sc.processNextWorkItem() {
+	}
+}
+
+func (sc *SnapshotController) processNextWorkItem() bool {
+	key, quit := sc.queue.Get()
+	if quit {
+		return false
+	}
+	defer sc.queue.Done(key)
+	err := sc.syncHandler(key.(string))
+	sc.handlerErr(err, key)
+	return true
+}
+
+func (sc *SnapshotController) handlerErr(err error, key interface{}) {
+	if err == nil {
+		sc.queue.Forget(key)
+		return
+	}
+
+	sc.logger.WithError(err).Warnf("Error syncing Longhorn snapshot %v", key)
+	sc.queue.AddRateLimited(key)
+	return
+
+	//if sc.queue.NumRequeues(key) < maxRetries {
+	//	sc.logger.WithError(err).Warnf("Error syncing Longhorn snapshot %v", key)
+	//	sc.queue.AddRateLimited(key)
+	//	return
+	//}
+
+	//utilruntime.HandleError(err)
+	//sc.logger.WithError(err).Warnf("Dropping Longhorn snapshot %v out of the queue", key)
+	//sc.queue.Forget(key)
```

Comment:
Do we need the retires?

Reply:
No. I think we don't need maxRetries. If there is any error, we add it back to the queue with rate limited to recheck it later.